### PR TITLE
Configuration

### PR DESCRIPTION
**Pull request by Kr0nox**
This draft is used as a cheklist for what needs to be implemented. @loernwerk/active please check of what is already done

- [x] router
- [x] controller: changing and reading configs
- [x] UI for changing configs
- [x] config rest interface
- [x] Maximale Sequenz Zahl frontend
- [x] Maximale Sequenz Zahl backend
- [x] Maximale Slide Zahl frontend
- [x] Maximale Slide Zahl backend

### CODE REVIEW

**Review by a-bumblebee — APPROVED** [review]

File: backend/controller/ConfigController.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * Manages configuration data in the database and handles requests for config requests
+ */
+export class ConfigController {
+    /**
+     * Searches a configurationentry from the database
+     * @param key the configurationkey of the entry
+     * @returns the value of the entry
+     */
+    public static async getConfigEntry(key: ConfigKey): Promise<unknown> {
+        return await DBConfigEntry.findOneBy({ key: key });
```
Comment (c0derMo) [inline backend/controller/ConfigController.ts]: This should check whether the config key exists - even if that's guaranteed by ensureConfig, I feel like this should throw an error if no config entry can be found

File: backend/controller/ConfigController.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    ): Promise<void> {
+        const entry = await DBConfigEntry.findOneBy({ key: key });
+        if (entry === null) {
+            throw new LoernwerkError(
+                'Config not Found',
+                LoernwerkErrorCodes.NOT_FOUND
+            );
+        }
+        entry.value = value;
+        entry.save();
```
Comment (c0derMo) [inline backend/controller/ConfigController.ts]: missing await

File: backend/controller/ConfigController.ts (hunk 1 above)
Comment (c0derMo) [reply backend/controller/ConfigController.ts]: Also, this should only return the value, not the entire DBConfigEntry

File: backend/controller/SequenceController.ts [hunk 3]
```diff
@@ -33,6 +35,19 @@ export class SequenceController {
...
             );
         }
+        const userSequenceLimit = (await ConfigController.getConfigEntry(
+            ConfigKey.MAX_SEQUENCES_PER_USER
+        )) as number;
+        const sequencesOfUser = await DBSequence.find({
+            where: { authorId: userId },
+            select: { code: true },
+        });
+        if (sequencesOfUser.length >= userSequenceLimit) {
```
Comment (c0derMo) [inline backend/controller/SequenceController.ts]: This also needs to check if sequenceUserLimit > 0

File: backend/controller/SequenceController.ts [hunk 4]
```diff
@@ -116,6 +131,20 @@ export class SequenceController {
             );
         }
 
+        const userSlideLimit = (await ConfigController.getConfigEntry(
+            ConfigKey.MAX_SLIDES_PER_SEQUENCE
+        )) as number;
+        const sequenceSlideCount = sequence.slides?.length;
+        if (
+            sequenceSlideCount !== undefined &&
+            sequenceSlideCount >= userSlideLimit
```
Comment (c0derMo) [inline backend/controller/SequenceController.ts:144]: This also needs to check if userSlideLimit > 0

File: backend/router/ConfigRouterFactory.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                await ConfigController.getAllConfigEntries();
+            res.status(200).json(allConfigEntries);
+        });
+
+        configRouter.get('/:key', requireLogin, async (req, res) => {
+            try {
+                const value = await ConfigController.getConfigEntry(
+                    req.params.key as ConfigKey
+                );
+                res.status(200).json(value);
```
Comment (c0derMo) [inline backend/router/ConfigRouterFactory.ts]: This should send `{ value }`

File: frontend/src/components/admin/ConfigEditor.vue [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+<script setup lang="ts">
+import { ConfigKey } from '../../../../model/configuration/ConfigKey';
+import { IConfigEntry } from '../../../../model/configuration/IConfigEntry';
+import { ConfigTypeMap } from '../../../../model/configuration/ConfigTypeMap';
+import { PropType, ref } from 'vue';
+import ButtonComponent from '../ButtonComponent.vue';
+import TextInputComponent from '../TextInputComponent.vue';
+
+const props = defineProps({
+  config: {
```
Comment (c0derMo) [inline frontend/src/components/admin/ConfigEditor.vue]: Per design, this is called `entries`

File: frontend/src/components/admin/ConfigEditor.vue [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  const type = ConfigTypeMap.getType(key);
+  switch (type.type) {
+    case 'number':
+      if (!(model.value[key] + '').match(/^[0-9]*$/)) {
+        return false;
+      }
+      return !(
+        type.options &&
+        type.options.length > 0 &&
+        type.options[0] == 'limited' &&
```
Comment (c0derMo) [inline frontend/src/components/admin/ConfigEditor.vue]: What is this for exactly?

File: model/configuration/IConfigType.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+     */
+    type: 'string' | 'number' | 'boolean' | 'enum';
+
+    /**
+     * If "type" is an enum, this variable saves valid options of the enum
+     */
+    options?: string[];
+}
+
+export type NumberOption = 'limited' | 'unlimited';
```
Comment (c0derMo) [inline model/configuration/IConfigType.ts]: Where is this used?

**Review by c0derMo — CHANGES_REQUESTED** [review]

**Review by ProgramPhoenix — APPROVED** [review]

File: frontend/src/components/admin/ConfigEditor.vue (hunk 7 above)
Comment (c0derMo) [reply frontend/src/components/admin/ConfigEditor.vue]: @Kr0nox since I think you wrote this

**Review by c0derMo — APPROVED** [review]

**Review by ProgramPhoenix — APPROVED** [review]

**Pull request merged**